### ISSUE TITLE
vomsapi is not thread-safe and crashes in a MT environment

### ISSUE THREAD

**Issue by abh3** [italiangrid/voms#89]
Just got a crash due to ssl_library_init() being called from two threads at the same time. A clear violation of openssl rules and guaranteed to crash. The reason is that the test for one-time initialization is not thread-safe. It's done this way

096 static bool initialized ;

121 if (!initialized) {
122     initialized = true; 

These are not atomic operations. Either declare initialized as std::atomic (which might not be possible depending on the compiler being used) or use GNU atomic extension on that variable. Alternatively, document how to force the initialization to happen in the main thread without doing any kinds of voms requests because there aren't any yet. Given how this is written, I suspect you expected that this would somehow happen. Though that is still problematic (see below).

But, this is only part of the problem. The voms code does not live in isolation. Other co-existing code (especially https and xroots) would have already called ssl_library_init() and it should not be called again in this case. There needs to be a way to tell the voms library to skip the openssl initialization and simply do what is presumably a one-time voms-specific initialization. That is, only call

 (void)AC_Init();
 InitProxyCertInfoExtension(1);

This then could be done by code that will be using voms after openssl was externally initialized but before everyone starts using openssl.

Finally, the code should be adapted to work with openssl prior to 1.1.0 and post 1.1.0 as the whole initialization structure has changed in that transition. This is documented here: https://wiki.openssl.org/index.php/Library_Initialization

**Comment by giacomini** [comment]
Thanks for the report.

> 096 static bool initialized ;
> 
> 121 if (!initialized) {
> 122 initialized = true;
> 
> These are not atomic operations. Either declare initialized as std::atomic (which might not be possible depending on the compiler being used) or use GNU atomic extension on that variable.

That's not enough. The whole initialization has to be atomic, not just the flag. We should use pthread_once or similar.

Said that, I don't think that the voms libs have been designed with MT in mind. We understand the need of use in an MT environment, but we need to assess the best way to achieve it.

**Comment by abh3** [comment]
I really think the problem is not at all complicated. From my code look and practical experience, voms is pretty much MT safe except for initialization. Now, if you could give me an API to skip the libssl initialization because it was already done; I could force the rest of initialization to simply occur in the main thread where MT issue don't exist. Note that if you adopt the recommended code for 1.1.0 I won't even need to use that API because libssl will automatically initialize itself only once (but that doesn't happen until 1.1.0 so I still would need it for older releases). So, really this doesn't require all that much assessment.

**Comment by giacomini** [comment]
I was thinking specifically to global/static mutable variables.

**Comment by giacomini** [comment]
I have committed a minimal change to address the OpenSSL initialization, or lack thereof.
@abh3 Let me know if the API is ok with you or any other comment you have.
Packages should be available shortly, in case you wish to test.

**Comment by abh3** [comment]
Hi Francesco,

I am looking at coms master at

https://github.com/italiangrid/voms/tree/master/src/api/ccapi

and I don't see any relevant updates. Am I looking in the correct place? 
If not, can you point me to the correct place and perhaps provide patch 
hashes?

Andy

**Comment by giacomini** [comment]
We start developments from the `develop` branch, from which we start feature/bug-specific branches. The commit is 7c60bd7 and is applied in branch `issue-89`. There is already a PR (#94). If you have comments, please add them to the PR. Note that Andrea has already proposed a change.
Both the commit and the PR are linked to this issue.

**Comment by abh3** [comment]
Thanks. I just added a single comment.

**Comment by giacomini** [comment]
@abh3 I've created PR #95 for the initialization of OpenSSL, if you wish to have a look.

**Comment by abh3** [comment]
Looks fine to me.

Andy
